File: frontend/src/features/commercial-archive/types/archive.ts
export type ArchiveSection = "archived" | "in_production" | "completed";
export type ArchiveFileKind = "pdf" | "xlsx" | "schema" | "xlsx_delivery_in_unit";
export type ProductType =
  | "plates"
  | "piles"
  | "steps"
  | "marches"
  | "bridge_piles"
  | "composite_piles"
  | "fbs"
  | "mixed";
export type ArchiveProductTypeFilter =
  | "all"
  | "plates"
  | "piles"
  | "steps"
  | "marches"
  | "bridge_piles"
  | "composite_piles"
  | "fbs";

export type ArchiveOfferListItem = {
  kp_id: number;
  creation_date: string | null;
  customer_name: string | null;
  manager_name: string | null;
  discount_percent: number;
  subtotal: number;
  vat_amount: number;
  total_amount: number;
  execution_terms: string | null;
  status: string | null;
  completion_percentage: number | null;
  sgp_progress?: { n: number; m: number } | null;
  /** Отгружено рейсами «обработано»: x из m (m = ordered_qty КП). */
  shipped_progress?: { x: number; m: number } | null;
  product_type?: ProductType;
  /** Concrete types for multi badges (Q3); preferred over a single mixed product_type. */
  product_types?: ProductType[];
  counterparty_id?: number | null;
  /** Optional: backend may omit — list badge skipped when absent. */
  has_delivery_schedule?: boolean;
};

export type ArchivePlateItem = {
  /** kp_plates.id — нужен для графика поставки; может отсутствовать у агрегатов. */
  id?: number | null;
  position_number: number | null;
  plate_name: string;
  length_m: number | null;
  width_m: number | null;
  load_class: number | null;
  qty: number;
  unit_price: number | null;
  discounted_price: number | null;
  unit_weight: number | null;
  total_weight: number | null;
  status: string | null;
  frost_resistance?: string | null;
  waterproofness?: string | null;
  concrete_aggregate?: string | null;
  concrete_spec_source?: string | null;
};

export type ArchivePileItem = {
  position_number: number | null;
  mark: string;
  concrete_grade: string;
  qty: number;
  unit_price: number | null;
  discounted_price: number | null;
  frost_resistance?: string | null;
  waterproofness?: string | null;
  concrete_aggregate?: string | null;
  concrete_spec_source?: string | null;
};

export type ArchiveStepItem = {
  position_number: number | null;
  mark: string;
  qty: number;
  unit_price: number | null;
  discounted_price: number | null;
};

export type ArchiveMarchItem = {
  position_number: number | null;
  mark: string;
  concrete_grade: string;
  qty: number;
  unit_price: number | null;
  discounted_price: number | null;
  frost_resistance?: string | null;
  waterproofness?: string | null;
  concrete_aggregate?: string | null;
  concrete_spec_source?: string | null;
};

export type ArchiveBridgePileItem = {
  position_number: number | null;
  mark: string;
  concrete_grade: string;
  qty: number;
  unit_price: number | null;
  discounted_price: number | null;
  frost_resistance?: string | null;
  waterproofness?: string | null;
  concrete_aggregate?: string | null;
  concrete_spec_source?: string | null;
};

export type ArchiveOfferFinance = {
  subtotal: number;
  vat_amount: number;
  total_amount: number;
  discount_percent: number;
};

export type SgpProgress = {
  n: number;
  m: number;
};

export type KpReadinessStepState = "done" | "active" | "pending" | "disabled";

export type KpReadinessStep = {
  id: "kp" | "production" | "sgp" | "release" | "closed";
  label: string;
  state: KpReadinessStepState;
  hint?: string | null;
};

export type KpReadinessSummary = {
  completion_percentage: number | null;
  sgp_progress: SgpProgress | null;
  issuable_qty: number;
  in_production_qty: number;
  summary_text: string;
  client_copy_text: string;
  steps: KpReadinessStep[];
  release_note?: string | null;
  expected_sgp_date?: string | null;
  expected_sgp_date_label?: string | null;
  fully_scheduled?: boolean;
};

export type KpReadinessPositionItem = {
  position_number: number | null;
  plate_name: string;
  length_m: number | null;
  width_m: number | null;
  load_class: number | null;
  label: string;
  ordered: number;
  in_plan: number;
  on_sgp: number;
  remaining: number;
};

export type KpReadinessPositionsResponse = {
  items: KpReadinessPositionItem[];
  count: number;
};

export type ArchiveOfferDetails = {
  kp_id: number;
  creation_date: string | null;
  customer_name: string | null;
  customer_inn?: string | null;
  customer_kpp?: string | null;
  counterparty_id?: number | null;
  manager_name: string | null;
  status: string | null;
  execution_terms: string | null;
  delivery_conditions: string | null;
  payment_conditions: string | null;
  finance: ArchiveOfferFinance;
  /** Стоимость одного рейса — то же поле, что logistics_cost при создании КП. */
  logistics_cost: number;
  pile_logistics_cost?: number;
  pile_trip_overrides?: Record<string, number>;
  pile_trips?: number;
  pile_trip_pending_marks?: string[];
  pile_delivery_ready?: boolean;
  plate_delivery_total?: number;
  pile_delivery_total?: number;
  fbs_lm_delivery_total?: number;
  fbs_lm_cargo_kg?: number;
  fbs_lm_trips?: number;
  fbs_lm_delivery_ready?: boolean;
  fbs_lm_pending_marks?: string[];
  fbs_lm_delivery_enabled?: boolean;
  long_pile_delivery_enabled?: boolean;
  long_pile_delivery_total?: number;
  long_pile_lengths?: Array<{
    length_key: number;
    trips: number;
    trip_cost: number | null;
    pending_marks: string[];
    ready: boolean;
    amount: number;
    qty: number;
  }>;
  long_pile_pending_marks?: string[];
  /** Масса груза (кг) по тем же правилам, что PDF/XLSX (resolve_kp_line_weight_kg на бэкенде). */
  total_cargo_weight_kg: number;
  /** Строка «Услуга по доставке грузов» = logistics_cost × число рейсов. */
  delivery_service_total_rub: number;
  product_type?: ProductType;
  plates: ArchivePlateItem[];
  piles?: ArchivePileItem[];
  steps?: ArchiveStepItem[];
  marches?: ArchiveMarchItem[];
  bridge_piles?: ArchiveBridgePileItem[];
  composite_piles?: ArchiveBridgePileItem[];
  fbs?: ArchiveBridgePileItem[];
  completion_percentage: number | null;
  readiness?: KpReadinessSummary | null;
};

export type ArchiveSearchState =
  | { kind: "number"; value: number }
  | { kind: "customer"; value: string }
  | null;

export type ArchiveSearchResponse = {
  mode: "number" | "customer";
  items: ArchiveOfferListItem[];
  total: number;
  truncated: boolean;
};

/** Ответ /archive/search для admin/manager. */
export type ArchiveSearchApiResponse = ArchiveSearchResponse;

export type ProductionEstimate = {
  total_length_m: number;
  estimated_tracks: number;
  estimated_days: number;
};

File: app/schemas/archive.py
from __future__ import annotations

from enum import Enum
from typing import Literal

from pydantic import BaseModel, ConfigDict, Field

from app.schemas.sgp import SgpProgress


ArchiveSection = Literal["archived", "in_production", "completed"]
ArchiveFileKind = Literal["pdf", "xlsx", "schema"]
ProductType = Literal["plates", "piles", "steps", "marches", "bridge_piles", "fbs", "mixed"]
ArchiveProductTypeFilter = Literal["all", "plates", "piles", "steps", "marches", "bridge_piles"]


class ArchiveOfferListItem(BaseModel):
    """Строка списка архива (аналог inline-кнопки в боте)."""

    model_config = ConfigDict(populate_by_name=True)

    kp_id: int
    creation_date: str | None = None
    customer_name: str | None = None
    manager_name: str | None = None
    discount_percent: float = 0.0
    subtotal: float = 0.0
    vat_amount: float = 0.0
    total_amount: float = 0.0
    execution_terms: str | None = None
    status: str | None = None
    completion_percentage: float | None = Field(
        default=None,
        description="Процент выполнения плит (для разделов in_production / completed).",
    )
    sgp_progress: dict[str, int] | None = Field(
        default=None,
        description="Бейдж N/M на СГП: {n, m}.",
    )
    shipped_progress: dict[str, int] | None = Field(
        default=None,
        description="Бейдж отгрузки: {x, m}, x = отгружено плит по done-рейсам, m = ordered_qty.",
    )
    product_type: ProductType = "plates"
    product_types: list[str] = Field(
        default_factory=list,
        description="Типы номенклатуры в КП для бейджей (Q3: N badges; mixed → concrete types).",
    )


class ArchivePileItem(BaseModel):
    position_number: int | None = None
    mark: str = ""
    concrete_grade: str = ""
    qty: int = 0
    unit_price: float | None = None
    discounted_price: float | None = None


class ArchiveStepItem(BaseModel):
    position_number: int | None = None
    mark: str = ""
    qty: int = 0
    unit_price: float | None = None
    discounted_price: float | None = None


class ArchiveMarchItem(BaseModel):
    position_number: int | None = None
    mark: str = ""
    concrete_grade: str = ""
    qty: int = 0
    unit_price: float | None = None
    discounted_price: float | None = None


class ArchiveBridgePileItem(BaseModel):
    position_number: int | None = None
    mark: str = ""
    concrete_grade: str = ""
    qty: int = 0
    unit_price: float | None = None
    discounted_price: float | None = None



class ArchiveFbsItem(BaseModel):
    position_number: int | None = None
    mark: str = ""
    concrete_grade: str = ""
    qty: int = 0
    unit_price: float | None = None
    discounted_price: float | None = None


class ArchivePlateItem(BaseModel):
    id: int | None = None  # kp_plates.id — нужен для графика поставки
    position_number: int | None = None
    plate_name: str = ""
    length_m: float | None = None
    width_m: float | None = None
    load_class: int | None = None
    qty: int = 0
    unit_price: float | None = None
    discounted_price: float | None = None
    unit_weight: float | None = None
    total_weight: float | None = None
    status: str | None = None


class ArchiveOfferFinance(BaseModel):
    subtotal: float = 0.0
    vat_amount: float = 0.0
    total_amount: float = 0.0
    discount_percent: float = 0.0


class KpReadinessStepState(str, Enum):
    DONE = "done"
    ACTIVE = "active"
    PENDING = "pending"
    DISABLED = "disabled"


class KpReadinessStep(BaseModel):
    id: Literal["kp", "production", "sgp", "release", "closed"]
    label: str
    state: KpReadinessStepState
    hint: str | None = None


class KpReadinessSummary(BaseModel):
    completion_percentage: float | None = None
    sgp_progress: SgpProgress | None = None
    issuable_qty: int = 0
    in_production_qty: int = 0
    summary_text: str = ""
    client_copy_text: str = ""
    steps: list[KpReadinessStep] = Field(default_factory=list)
    release_note: str | None = None
    expected_sgp_date: str | None = Field(
        default=None,
        description="ISO date YYYY-MM-DD; last planned production day",
    )
    expected_sgp_date_label: str | None = Field(
        default=None,
        description="Formatted DD.MM.YYYY for UI",
    )
    fully_scheduled: bool = False


class KpReadinessPositionItem(BaseModel):
    position_number: int | None = None
    plate_name: str
    length_m: float | None = None
    width_m: float | None = None
    load_class: int | None = None
    label: str
    ordered: int
    in_plan: int
    on_sgp: int
    remaining: int


class KpReadinessPositionsResponse(BaseModel):
    items: list[KpReadinessPositionItem]
    count: int


class ArchiveOfferDetails(BaseModel):
    """Полная карточка КП для страницы архива."""

    kp_id: int
    creation_date: str | None = None
    customer_name: str | None = None
    manager_name: str | None = None
    status: str | None = None
    execution_terms: str | None = None
    delivery_conditions: str | None = None
    payment_conditions: str | None = None
    finance: ArchiveOfferFinance
    logistics_cost: float = Field(default=0.0, description="Стоимость одного рейса (как logistics_cost при создании КП).")
    pile_logistics_cost: float = Field(default=0.0, description="Тариф рейса свай/мостовых.")
    pile_trip_overrides: dict[str, int] = Field(default_factory=dict)
    pile_trips: int = 0
    pile_trip_pending_marks: list[str] = Field(default_factory=list)
    pile_delivery_ready: bool = True
    plate_delivery_total: float = 0.0
    pile_delivery_total: float = 0.0
    total_cargo_weight_kg: float = Field(default=0.0, description="Суммарная масса по строкам через resolve_kp_line_weight_kg (как PDF/XLSX).")
    delivery_service_total_rub: float = Field(
        default=0.0,
        description="Строка «Услуга по доставке грузов»: рейсы × стоимость рейса.",
    )
    product_type: ProductType = "plates"
    plates: list[ArchivePlateItem] = Field(default_factory=list)
    piles: list[ArchivePileItem] = Field(default_factory=list)
    steps: list[ArchiveStepItem] = Field(default_factory=list)
    marches: list[ArchiveMarchItem] = Field(default_factory=list)
    bridge_piles: list[ArchiveBridgePileItem] = Field(default_factory=list)
    fbs: list[ArchiveFbsItem] = Field(default_factory=list)
    completion_percentage: float | None = None
    readiness: KpReadinessSummary | None = None


class UpdateDiscountRequest(BaseModel):
    discount: float = Field(ge=0, le=100)


class UpdateLogisticsCostRequest(BaseModel):
    logistics_cost: float = Field(ge=0, description="Новая стоимость одного рейса плит.")
    pile_logistics_cost: float | None = Field(default=None, ge=0)
    pile_trip_overrides: dict[str, int] | None = None


class MoveToProductionRequest(BaseModel):
    execution_terms: str = Field(min_length=1, max_length=128)


class CapacityDayInfo(BaseModel):
    model_config = ConfigDict(populate_by_name=True)

    occupied: float = 0
    max: float = 5  # noqa: A003 — API field matches days_info.max


class CapacitySnapshotResponse(BaseModel):
    """Снимок ёмкости завода для виджета (move / график поставок)."""

    start_date: str
    target_date: str
    tracks_needed: int
    tracks_free_in_window: int
    delta: int
    status: Literal["green", "yellow", "red"]
    hint: str | None = None
    days_info: dict[str, CapacityDayInfo] = Field(default_factory=dict)
    holidays: list[str] = Field(default_factory=list)
    extra_workdays: list[str] = Field(default_factory=list)
    calendar_from_month: str
    calendar_to_month: str


class ArchiveSearchResponse(BaseModel):
    mode: Literal["number", "customer"]
    items: list[ArchiveOfferListItem] = Field(default_factory=list)
    total: int = 0
    truncated: bool = False
